File: native/cocos/renderer/renderer/BaseRenderer.cpp
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "BaseRenderer.h"
#include <new>
#include "gfx/DeviceGraphics.h"
#include "gfx/Texture2D.h"
#include "ProgramLib.h"
#include "View.h"
#include "Scene.h"
#include "Effect.h"
#include "Light.h"
#include "InputAssembler.h"
#include "Pass.h"
#include "Camera.h"
#include "Model.h"
#include "math/MathUtil.h"
#include "Program.h"

RENDERER_BEGIN

const size_t BaseRenderer::cc_dirLightDirection = std::hash<std::string>{}("cc_dirLightDirection");
const size_t BaseRenderer::cc_dirLightColor = std::hash<std::string>{}("cc_dirLightColor");
const size_t BaseRenderer::cc_pointLightPositionAndRange = std::hash<std::string>{}("cc_pointLightPositionAndRange");
const size_t BaseRenderer::cc_pointLightColor = std::hash<std::string>{}("cc_pointLightColor");
const size_t BaseRenderer::cc_spotLightDirection = std::hash<std::string>{}("cc_spotLightDirection");
const size_t BaseRenderer::cc_spotLightPositionAndRange = std::hash<std::string>{}("cc_spotLightPositionAndRange");
const size_t BaseRenderer::cc_spotLightColor = std::hash<std::string>{}("cc_spotLightColor");
const size_t BaseRenderer::cc_shadow_map = std::hash<std::string>{}("cc_shadow_map");
const size_t BaseRenderer::cc_shadow_map_lightViewProjMatrix = std::hash<std::string>{}("cc_shadow_map_lightViewProjMatrix");
const size_t BaseRenderer::cc_shadow_map_info = std::hash<std::string>{}("cc_shadow_map_info");
const size_t BaseRenderer::cc_shadow_map_bias = std::hash<std::string>{}("cc_shadow_map_bias");
const size_t BaseRenderer::cc_shadow_lightViewProjMatrix = std::hash<std::string>{}("cc_shadow_lightViewProjMatrix");
const size_t BaseRenderer::cc_shadow_info = std::hash<std::string>{}("cc_shadow_info");
const size_t BaseRenderer::cc_matView = std::hash<std::string>{}("cc_matView");
const size_t BaseRenderer::cc_matWorld = std::hash<std::string>{}("cc_matWorld");
const size_t BaseRenderer::cc_matWorldIT = std::hash<std::string>{}("cc_matWorldIT");
const size_t BaseRenderer::cc_matpProj = std::hash<std::string>{}("cc_matpProj");
const size_t BaseRenderer::cc_matViewProj = std::hash<std::string>{}("cc_matViewProj");
const size_t BaseRenderer::cc_cameraPos = std::hash<std::string>{}("cc_cameraPos");

BaseRenderer::BaseRenderer()
{
    _drawItems = new RecyclePool<DrawItem>([]()mutable->DrawItem*{return new DrawItem();},100);
    _stageInfos = new RecyclePool<StageInfo>([]()mutable->StageInfo*{return new StageInfo();}, 10);
    _views = new RecyclePool<View>([]()mutable->View*{return new View();}, 8);
    
    _tmpMat4 = new cocos2d::Mat4();
}

BaseRenderer::~BaseRenderer()
{
    _device->release();
    _device = nullptr;
    
    delete _programLib;
    _programLib = nullptr;
    
    RENDERER_SAFE_RELEASE(_defaultTexture);
    _defaultTexture = nullptr;
    
    delete _drawItems;
    _drawItems = nullptr;
    
    delete _stageInfos;
    _stageInfos = nullptr;
    
    delete _views;
    _views = nullptr;
    
    delete _tmpMat4;
    _tmpMat4 = nullptr;
}

bool BaseRenderer::init(DeviceGraphics* device, std::vector<ProgramLib::Template>& programTemplates)
{
    _device = device;
    _device->retain();
    _programLib = new (std::nothrow) ProgramLib(_device, programTemplates);
    return true;
}

bool BaseRenderer::init(DeviceGraphics* device, std::vector<ProgramLib::Template>& programTemplates, Texture2D* defaultTexture)
{
    _device = device;
    _device->retain();
    _defaultTexture = defaultTexture;
    RENDERER_SAFE_RETAIN(_defaultTexture);
    _programLib = new (std::nothrow) ProgramLib(_device, programTemplates);
    return true;
}

void BaseRenderer::registerStage(const std::string& name, const StageCallback& callback)
{
    _stage2fn.emplace(std::make_pair(name, callback));
}

// protected functions

void BaseRenderer::render(const View& view, const Scene* scene)
{
    // setup framebuffer
    _device->setFrameBuffer(view.frameBuffer);
    
    // setup viewport
    _device->setViewport(view.rect.x,
                         view.rect.y,
                         view.rect.w,
                         view.rect.h);
    
    // setup clear
    Color4F clearColor;
    if (ClearFlag::COLOR & view.clearFlags)
        clearColor = view.color;
    _device->clear(view.clearFlags, &clearColor, view.depth, view.stencil);
    
    // get all draw items
    _drawItems->reset();
    int modelMask = -1;
    for (const auto& model : scene->getModels())
    {
        modelMask = model->getCullingMask();
        if (view.cullingByID)
        {
            if ((modelMask & view.cullingMask) == 0)
                continue;
        }
        else
        {
            if (-1 != modelMask)
                continue;
        }
        
        DrawItem* drawItem = _drawItems->add();
        model->extractDrawItem(*drawItem);
    }
    
    // dispatch draw items to different stage
    _stageInfos->reset();
    StageItem stageItem;
    std::vector<StageItem> stageItems;
    for (const auto& stage : view.stages)
    {
        for (size_t i = 0, len = _drawItems->getLength(); i < len; i++)
        {
            const DrawItem* item = _drawItems->getData(i);
            auto tech = item->effect->getTechnique(stage);
            if (tech)
            {
                stageItem.model = item->model;
                stageItem.ia = item->ia;
                stageItem.effect = item->effect;
                stageItem.defines = item->defines;
                stageItem.technique = tech;
                stageItem.sortKey = -1;
                stageItem.uniforms = item->uniforms;
                stageItem.definesKeyHash = item->definesKeyHash;
                
                stageItems.push_back(stageItem);
            }
        }
        StageInfo* stageInfo = _stageInfos->add();
        stageInfo->stage = stage;
        stageInfo->items = &stageItems;
    }
    
    // render stages
    std::unordered_map<std::string, const StageCallback>::iterator foundIter;
    for (size_t i = 0, len = _stageInfos->getLength(); i < len; i++)
    {
        const StageInfo* stageInfo = _stageInfos->getData(i);
        foundIter = _stage2fn.find(stageInfo->stage);
        if (_stage2fn.end() != foundIter)
        {
            auto& fn = foundIter->second;
            fn(view, *stageInfo->items);
        }
    }
}

void BaseRenderer::setProperty (Effect::Property& prop)
{
    Technique::Parameter::Type propType = prop.getType();
    auto& propName = prop.getName();
    auto propHashName = prop.getHashName();
    if (Effect::Property::Type::UNKNOWN == propType)
    {
        RENDERER_LOGW("Failed to set technique property, type unknown");
        return;
    }
    
    if (nullptr == prop.getValue())
    {
        prop = Effect::Property(propName, propType);
        
        if (Effect::Property::Type::TEXTURE_2D == propType)
        {
            prop.setTexture(_defaultTexture);
        }
    }
    
    if (nullptr == prop.getValue())
    {
        RENDERER_LOGW("Failed to set technique property %s, value not found", propName.c_str());
        return;
    }
    
    if (Effect::Property::Type::TEXTURE_2D == propType ||
        Effect::Property::Type::TEXTURE_CUBE == propType)
    {
        if (1 == prop.getCount())
        {
            _device->setTexture(propHashName,
                                (renderer::Texture *)(prop.getValue()),
                                allocTextureUnit());
        }
        else if (0 < prop.getCount())
        {
            std::vector<int> slots;
            slots.reserve(10);
            for (int i = 0; i < prop.getCount(); ++i)
            {
                slots.push_back(allocTextureUnit());
            }
            
            _device->setTextureArray(propHashName,
                                     prop.getTextureArray(),
                                     slots);
        }
    }
    else
    {
        if (0 != prop.getCount())
        {
//            if (Technique::Parameter::Type::COLOR3 == propType ||
//                Technique::Parameter::Type::INT3 == propType ||
//                Technique::Parameter::Type::FLOAT3 == propType ||
//                Technique::Parameter::Type::MAT3 == propType)
//            {
//                RENDERER_LOGW("Uinform array of color3/int3/float3/mat3 can not be supported!");
//                return;
//            }
            
            uint8_t size = Technique::Parameter::getElements(propType);
            if (size * prop.getCount() > 64)
            {
                RENDERER_LOGW("Uniform array is too long!");
                return;
            }
        }
        
        uint16_t bytes = prop.getBytes();
        if (Effect::Property::Type::INT == propType ||
            Effect::Property::Type::INT2 == propType ||
            Effect::Property::Type::INT4 == propType)
        {
            _device->setUniformiv(propHashName, bytes / sizeof(int), (const int*)prop.getValue());
        }
        else
        {
            _device->setUniformfv(propHashName, bytes / sizeof(float), (const float*)prop.getValue());
        }
    }
}

void BaseRenderer::draw(const StageItem& item)
{
    const Mat4& worldMatrix = item.model->getWorldMatrix();
    _device->setUniformMat4(cc_matWorld, worldMatrix.m);
    
    _tmpMat4->set(worldMatrix);
    _tmpMat4->inverse();
    _tmpMat4->transpose();
    _device->setUniformMat4(cc_matWorldIT, _tmpMat4->m);
    
    // set technique uniforms
    for (int i = 0, len = (int)item.uniforms->size(); i < len; i++)
    {
        std::unordered_map<std::string, Effect::Property>* properties = (*item.uniforms)[i];
        for (auto& prop : *properties) {
            setProperty(prop.second);
        }
    }
    
    auto ia = item.ia;
    // for each pass
    for (const auto& pass : item.technique->getPasses())
    {
        // set vertex buffer
        _device->setVertexBuffer(0, ia->getVertexBuffer());
        
        // set index buffer
        if (ia->_indexBuffer)
            _device->setIndexBuffer(ia->_indexBuffer);
        
        // set primitive type
        _device->setPrimitiveType(ia->_primitiveType);
        
        // get program
        _program = _programLib->switchProgram(pass->getHashName(), item.definesKeyHash, *(item.defines));
        _device->setProgram(_program);
        
        // cull mode
        _device->setCullMode(pass->_cullMode);
        
        // blend
        if (pass->_blend)
        {
            _device->enableBlend();
            _device->setBlendFuncSeparate(pass->_blendSrc,
                                          pass->_blendDst,
                                          pass->_blendSrcAlpha,
                                          pass->_blendDstAlpha);
            _device->setBlendEquationSeparate(pass->_blendEq, pass->_blendAlphaEq);
            _device->setBlendColor(pass->_blendColor);
        }
        
        // depth test & write
        if (pass->_depthTest)
        {
            _device->enableDepthTest();
            _device->setDepthFunc(pass->_depthFunc);
        }
        if (pass->_depthWrite)
            _device->enableDepthWrite();
        
        // setencil
        if (pass->_stencilTest)
        {
            _device->enableStencilTest();
            
            // front
            _device->setStencilFuncFront(pass->_stencilFuncFront,
                                         pass->_stencilRefFront,
                                         pass->_stencilMaskFront);
            _device->setStencilOpFront(pass->_stencilFailOpFront,
                                       pass->_stencilZFailOpFront,
                                       pass->_stencilZPassOpFront,
                                       pass->_stencilWriteMaskFront);
            
            // back
            _device->setStencilFuncBack(pass->_stencilFuncBack,
                                        pass->_stencilRefBack,
                                        pass->_stencilMaskBack);
            _device->setStencilOpBack(pass->_stencilFailOpBack,
                                      pass->_stencilZFailOpBack,
                                      pass->_stencilZPassOpBack,
                                      pass->_stencilWriteMaskBack);
        }
        
        // draw pass
        _device->draw(ia->_start, ia->getPrimitiveCount());
        
        resetTextureUint();
    }
}

// private functions

void BaseRenderer::resetTextureUint()
{
    _usedTextureUnits = 0;
}

int BaseRenderer::allocTextureUnit()
{
    int maxTexureUnits = _device->getCapacity().maxTextureUnits;
    if (_usedTextureUnits >= maxTexureUnits)
        RENDERER_LOGW("Trying to use %d texture uints while this GPU only supports %d", _usedTextureUnits, maxTexureUnits);
    
    return _usedTextureUnits++;
}

void BaseRenderer::reset()
{
    _views->reset();
    _stageInfos->reset();
}

View* BaseRenderer::requestView()
{
    return _views->add();
}

RENDERER_END

File: cocos2d/renderer/core/base-renderer.js
// Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

import { RecyclePool } from '../memop';
import enums from '../enums';
import { Vec2, Vec4, Mat3, Mat4, Color, Vec3 } from '../../core/value-types';
import ProgramLib from './program-lib';
import View from './view';
import gfx from '../gfx';

let _m4_tmp = new Mat4();

let _stageInfos = new RecyclePool(() => {
  return {
    stage: null,
    items: null,
  };
}, 8);

let _float2_pool = new RecyclePool(() => {
  return new Float32Array(2);
}, 8);

let _float3_pool = new RecyclePool(() => {
  return new Float32Array(3);
}, 8);

let _float4_pool = new RecyclePool(() => {
  return new Float32Array(4);
}, 8);

let _float9_pool = new RecyclePool(() => {
  return new Float32Array(9);
}, 8);

let _float16_pool = new RecyclePool(() => {
  return new Float32Array(16);
}, 8);

let _float64_pool = new RecyclePool(() => {
  return new Float32Array(64);
}, 8);

let _int2_pool = new RecyclePool(() => {
  return new Int32Array(2);
}, 8);

let _int3_pool = new RecyclePool(() => {
  return new Int32Array(3);
}, 8);

let _int4_pool = new RecyclePool(() => {
  return new Int32Array(4);
}, 8);

let _int64_pool = new RecyclePool(() => {
  return new Int32Array(64);
}, 8);

export default class Base {
  /**
   * @param {gfx.Device} device
   * @param {Object} opts
   * @param {gfx.Texture2D} opts.defaultTexture
   * @param {gfx.TextureCube} opts.defaultTextureCube
   */
  constructor (device, opts) {
    this._device = device;
    this._programLib = new ProgramLib(device);
    this._opts = opts;
    this._type2defaultValue = {
      [enums.PARAM_INT]: 0,
      [enums.PARAM_INT2]: new Vec2(0, 0),
      [enums.PARAM_INT3]: new Vec3(0, 0, 0),
      [enums.PARAM_INT4]: new Vec4(0, 0, 0, 0),
      [enums.PARAM_FLOAT]: 0.0,
      [enums.PARAM_FLOAT2]: new Vec2(0, 0),
      [enums.PARAM_FLOAT3]: new Vec3(0, 0, 0),
      [enums.PARAM_FLOAT4]: new Vec4(0, 0, 0, 0),
      [enums.PARAM_COLOR4]: new Color(0, 0, 0, 1),
      [enums.PARAM_MAT3]: new Mat3(),
      [enums.PARAM_MAT4]: new Mat4(),
      [enums.PARAM_TEXTURE_2D]: opts.defaultTexture,
      [enums.PARAM_TEXTURE_CUBE]: opts.defaultTextureCube,
    };
    this._stage2fn = {};
    this._usedTextureUnits = 0;

    this._viewPools = new RecyclePool(() => {
      return new View();
    }, 8);

    this._drawItemsPools = new RecyclePool(() => {
      return {
        model: null,
        node: null,
        ia: null,
        effect: null,
        defines: null,
        uniforms: null
      };
    }, 100);

    this._stageItemsPools = new RecyclePool(() => {
      return new RecyclePool(() => {
        return {
          model: null,
          node: null,
          ia: null,
          effect: null,
          defines: null,
          passes: [],
          sortKey: -1,
          uniforms: null
        };
      }, 100);
    }, 16);

    this._definesChanged = false;
  }

  _resetTextuerUnit () {
    this._usedTextureUnits = 0;
  }

  _allocTextureUnit () {
    const device = this._device;

    let unit = this._usedTextureUnits;
    if (unit >= device._caps.maxTextureUnits) {
      console.warn(`Trying to use ${unit} texture units while this GPU supports only ${device._caps.maxTextureUnits}`);
    }

    this._usedTextureUnits += 1;
    return unit;
  }

  _registerStage (name, fn) {
    this._stage2fn[name] = fn;
  }

  clear () {
    this._programLib.clear();
    this.reset();
  }

  reset () {
    this._viewPools.reset();
    this._stageItemsPools.reset();

    this._definesChanged = false;
  }

  _requestView () {
    return this._viewPools.add();
  }

  _render (view, scene) {
    const device = this._device;

    // setup framebuffer
    device.setFrameBuffer(view._framebuffer);

    // setup viewport
    device.setViewport(
      view._rect.x,
      view._rect.y,
      view._rect.w,
      view._rect.h
    );

    // setup clear
    let clearOpts = {};
    if (view._clearFlags & enums.CLEAR_COLOR) {
      clearOpts.color = Vec4.toArray([], view._color);
    }
    if (view._clearFlags & enums.CLEAR_DEPTH) {
      clearOpts.depth = view._depth;
    }
    if (view._clearFlags & enums.CLEAR_STENCIL) {
      clearOpts.stencil = view._stencil;
    }
    device.clear(clearOpts);

    // get all draw items
    this._drawItemsPools.reset();

    for (let i = 0; i < scene._models.length; ++i) {
      let model = scene._models.data[i];

      // filter model by view
      if ((model._cullingMask & view._cullingMask) === 0) {
        continue;
      }

      let drawItem = this._drawItemsPools.add();
      model.extractDrawItem(drawItem);
    }

    // dispatch draw items to different stage
    _stageInfos.reset();

    for (let i = 0; i < view._stages.length; ++i) {
      let stage = view._stages[i];
      let stageItems = this._stageItemsPools.add();
      stageItems.reset();

      for (let j = 0; j < this._drawItemsPools.length; ++j) {
        let drawItem = this._drawItemsPools.data[j];
        let passes = drawItem.effect.stagePasses[stage];
        if (!passes || passes.length === 0) continue;

        let stageItem = stageItems.add();
        stageItem.passes = passes;
        stageItem.model = drawItem.model;
        stageItem.node = drawItem.node;
        stageItem.ia = drawItem.ia;
        stageItem.effect = drawItem.effect;
        stageItem.defines = drawItem.defines;
        stageItem.sortKey = -1;
        stageItem.uniforms = drawItem.uniforms;
      }

      let stageInfo = _stageInfos.add();
      stageInfo.stage = stage;
      stageInfo.items = stageItems;
    }

    // render stages
    for (let i = 0; i < _stageInfos.length; ++i) {
      let info = _stageInfos.data[i];
      let fn = this._stage2fn[info.stage];
      fn(view, info.items);
    }
  }

  _setProperty (prop) {
    const device = this._device;
    let param = prop.value;

    if (param === undefined) {
      param = prop.val;
    }

    if (param === undefined) {
      param = this._type2defaultValue[prop.type];
    }

    if (param === undefined) {
      console.warn(`Failed to set technique property ${prop.name}, value not found.`);
      return;
    }

    if (
      prop.type === enums.PARAM_TEXTURE_2D ||
      prop.type === enums.PARAM_TEXTURE_CUBE
    ) {
      if (Array.isArray(param)) {
        if (param.length > prop.count) {
          console.error(`Failed to set property [${prop.name}] : The length of texture array [${param.length}] is bigger than [${prop.count}].`);
          return;
        }
        let slots = _int64_pool.add();
        for (let index = 0; index < param.length; ++index) {
          slots[index] = this._allocTextureUnit();
        }
        device.setTextureArray(prop.name, param, slots);
      } else {
        device.setTexture(prop.name, param, this._allocTextureUnit());
      }
    } else {
      if (prop.directly) {
        device.setUniformDirectly(prop.name, param);
      }
      else {
        device.setUniform(prop.name, param);
      }
    }
  }

  _draw (item) {
    const device = this._device;
    const programLib = this._programLib;
    const { node, ia, passes, effect } = item;

    // reset the pool
    // NOTE: we can use drawCounter optimize this
    // TODO: should be configurable
    _float2_pool.reset();
    _float3_pool.reset();
    _float4_pool.reset();
    _float9_pool.reset();
    _float16_pool.reset();
    _float64_pool.reset();
    _int2_pool.reset();
    _int3_pool.reset();
    _int4_pool.reset();
    _int64_pool.reset();

    // set common uniforms
    // TODO: try commit this depends on effect
    // {
    node.getWorldMatrix(_m4_tmp);
    device.setUniform('cc_matWorld', Mat4.toArray(_float16_pool.add(), _m4_tmp));

    // let wq = node.getWorldRotation(cc.quat());
    Mat4.invert(_m4_tmp, _m4_tmp);
    Mat4.transpose(_m4_tmp, _m4_tmp);
    device.setUniform('cc_matWorldIT', Mat4.toArray(_float16_pool.add(), _m4_tmp));
    // }

    let defines = this._defines;

    // for each pass
    for (let i = 0; i < passes.length; ++i) {
      let pass = passes[i];

      if (this._definesChanged) {
        pass._programKey = null;
      }

      let count = ia.count;

      // set vertex buffer
      if (ia._vertexBuffer) {
        device.setVertexBuffer(0, ia._vertexBuffer);
      }

      // set index buffer
      if (ia._indexBuffer) {
        device.setIndexBuffer(ia._indexBuffer);
      }

      // set primitive type
      device.setPrimitiveType(ia._primitiveType);

      // set program
      Object.setPrototypeOf(defines, pass._defines);

      let program = programLib.getProgram(pass, defines, effect.name);
      device.setProgram(program);

      let uniforms = program._uniforms;
      let variants = pass._properties;
      for (let j = 0; j < uniforms.length; j++) {
        let prop = variants[uniforms[j].name];
        if (prop !== undefined)
        this._setProperty(prop);
      }


      // cull mode
      device.setCullMode(pass._cullMode);

      // blend
      if (pass._blend) {
        device.enableBlend();
        device.setBlendFuncSep(
          pass._blendSrc,
          pass._blendDst,
          pass._blendSrcAlpha,
          pass._blendDstAlpha
        );
        device.setBlendEqSep(
          pass._blendEq,
          pass._blendAlphaEq
        );
        device.setBlendColor32(pass._blendColor);
      }

      // depth test & write
      if (pass._depthTest) {
        device.enableDepthTest();
        device.setDepthFunc(pass._depthFunc);
      }
      if (pass._depthWrite) {
        device.enableDepthWrite();
      }

      // stencil
      device.setStencilTest(pass._stencilTest);
      if (pass._stencilTest === gfx.STENCIL_ENABLE) {
        // front
        device.setStencilFuncFront(
          pass._stencilFuncFront,
          pass._stencilRefFront,
          pass._stencilMaskFront
        );
        device.setStencilOpFront(
          pass._stencilFailOpFront,
          pass._stencilZFailOpFront,
          pass._stencilZPassOpFront,
          pass._stencilWriteMaskFront
        );

        // back
        device.setStencilFuncBack(
          pass._stencilFuncBack,
          pass._stencilRefBack,
          pass._stencilMaskBack
        );
        device.setStencilOpBack(
          pass._stencilFailOpBack,
          pass._stencilZFailOpBack,
          pass._stencilZPassOpBack,
          pass._stencilWriteMaskBack
        );
      }

      // draw pass
      device.draw(ia._start, count);

      this._resetTextuerUnit();
    }
  }
}
